smolyay/samples.py

```diff
@@ -508,3 +508,620 @@ def _create(self):
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            for ps,d in zip(self.point_sets,domain):
+                ps.domain = d
+            self._valid_cache = False
+
+    @property
+    def point_sets(self):
+        """:class:UnidimensionalPointSet: set of unique points"""
+        return self._point_sets
+
+    @property
+    def indexes(self):
+        """list: Grid points' indexes."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._indexes
+
+
+class TensorProductPointSet(PointSetProduct):
+    """Generate a grid using all combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        self._points = numpy.array(list(itertools.product(*self._point_sets)))
+
+
+class SmolyakSparseProductPointSet(PointSetProduct):
+    """Generate a grid using sparse combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        grid_points = None
+        max_num_level = max([ob.num_levels for ob in self._point_sets])
+        max_num_levels = [ob.num_levels for ob in self._point_sets]
+        # get the combinations of levels
+        index_composition = []
+        for sum_of_levels in range(
+            self.num_dimensions, self.num_dimensions + max_num_level
+        ):
+            index_composition.extend(
+                list(
+                    generate_compositions(
+                        sum_of_levels, self.num_dimensions, include_zero=False
+                    )
+                )
+            )
+        index_composition = numpy.array(index_composition) - 1
+        if min(max_num_levels) != max_num_level:
+            # only check if point sets have different number of levels
+            valid_comb = numpy.all(
+                numpy.greater_equal(max_num_levels, index_composition), axis=1
+            )
+            index_composition = index_composition[valid_comb, :]
```

Comment:
What is this doing? Some commenting, etc. would be helpful.

Reply:
Basically, when generating the combinations of levels for each dimension, it generates them first assuming that all dimensions can take the same number of levels. This part of the code first checks if that's actually true, and if it isn't, it finds all the valid combinations and only keeps those. Valid combinations are ones where the level for every dimension doesn't exceed the number of levels for that dimension.

I have added comments and cleaned up this section of code.